internal/service/server/server.go

```diff
@@ -440,6 +454,58 @@ func ResourceServer() *schema.Resource {
 	}
 }
 
+func resourceServerStateUpgradeV0(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+	svc := meta.(*service.Service)
+	networking, err := svc.GetServerNetworks(ctx, &request.GetServerNetworksRequest{ServerUUID: rawState["id"].(string)})
+	if err != nil {
+		return rawState, err
+	}
+
+	if networking == nil {
+		return rawState, nil
+	}
+
+	rawInterfaces, ok := rawState["network_interface"].([]interface{})
+	if !ok {
+		return rawState, fmt.Errorf("network_interface not found in state")
+	}
+
+	if len(networking.Interfaces) != len(rawInterfaces) {
+		return rawState, fmt.Errorf("values for network_interface have been modified outside of Terraform, unable to migrate the state. correct the drift between state and the resource to continue")
+	}
+
+	for k, v := range rawInterfaces {
+		iface := v.(map[string]interface{})
+		index, err := getIndexFromNetworking(networking, iface)
+		if err != nil {
+			return rawState, err
+		}
+
+		iface["index"] = index
+		rawInterfaces[k] = iface
+	}
+
+	rawState["network_interface"] = rawInterfaces
+
+	return rawState, nil
+}
+
+func getIndexFromNetworking(networking *upcloud.Networking, iface map[string]interface{}) (int, error) {
+	for _, n := range networking.Interfaces {
+		if n.Type != iface["type"].(string) {
+			continue
+		}
+
+		if n.MAC != iface["mac_address"].(string) {
+			continue
+		}
+
+		return n.Index, nil
```

Comment:
Is this same as:

```suggestion
		if n.Type == iface["type"].(string) && n.MAC == iface["mac_address"].(string) {
			return n.Index, nil
		}
```

No clear preference on which one would be better, though